pkg/cluster/gossip.go

```diff
@@ -0,0 +1,253 @@
+//
+// GossipNode operations are unavailable until the node is started.
+func NewGossipNode(l log.Logger, srv *grpc.Server, c *GossipConfig) (*GossipNode, error) {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	sharder := shard.Ring(256)
+
+	ckitConfig := ckit.Config{
+		Name:          c.NodeName,
+		AdvertiseAddr: c.AdvertiseAddr,
+		Sharder:       sharder,
+		Log:           l,
+		Pool:          c.Pool,
+	}
+
+	inner, err := ckit.NewNode(srv, ckitConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GossipNode{
+		cfg:       c,
+		innerNode: inner,
+		log:       l,
+		sharder:   sharder,
+	}, nil
+}
+
+// ChangeState changes the state of n. ChangeState will block until the state
+// change has been receieved by another node; cancel the context to stop
+// waiting.
+//
+// Nodes must be a StateParticipant to be returned during Lookup.
+//
+// The following state transitions are valid:
+//
+// 		StatePending -> StateParticipant|StateViewer|StateTerminating|StateGone
+// 		StateParticipant -> StateTerminating|StateGone
+// 		StateViewer -> StateTerminating|StateGone
+// 		StateTerminating -> StateGone
+func (n *GossipNode) ChangeState(ctx context.Context, to peer.State) error {
+	if !n.started.Load() {
+		return fmt.Errorf("node not started")
+	}
+	return n.innerNode.ChangeState(ctx, to)
+}
+
+// CurrentState returns the current state of the node. Note that other nodes
+// may have an older view of the state while a state change propagates
+// throughout the cluster.
+func (n *GossipNode) CurrentState() peer.State {
+	return n.innerNode.CurrentState()
+}
+
+// Lookup implements Node and returns numOwners Peers that are responsible for
+// key. Only peers in StateParticipant are considered during a lookup; if no
+// peers are in StateParticipant, the Lookup will fail.
+func (n *GossipNode) Lookup(key shard.Key, numOwners int, op shard.Op) ([]peer.Peer, error) {
+	if !n.started.Load() {
+		return nil, fmt.Errorf("node not started")
+	}
+	return n.sharder.Lookup(key, numOwners, op)
+}
+
+// Observe registers o to be informed when the cluster changes, including peers
+// appearing, disapearing, or changing state.
+//
+// Calls will have to filter events if they are only interested in a subset of
+// changes.
+func (n *GossipNode) Observe(o ckit.Observer) {
+	n.innerNode.Observe(o)
+}
+
+// Peers returns the current set of Peers.
+func (n *GossipNode) Peers() []peer.Peer {
+	return n.innerNode.Peers()
+}
+
+// Start starts the node. Start will connect to peers if configured to do so.
+//
+// Start must only be called after the gRPC server is running, otherwise Start
+// will block forever.
+func (n *GossipNode) Start() (err error) {
+	defer func() {
+		if err == nil {
+			n.started.Store(true)
+		}
+	}()
+	return n.innerNode.Start(n.cfg.JoinPeers)
+}
+
+// Stop leaves the cluster and terminates n. n cannot be re-used after
+// stopping.
+//
+// It is advisble to ChangeState to StateTerminating and StateGone before
```

Comment:
It might be worth considering to make nodes perform these state transitions themselves when shutting down

Reply:
👍 we'll probably end up doing this, we'll need to see once everything is wired in if we can automate the shutdown transitions and stop exposing this code